H4HIP: Helm Sequencing Proposal (#373)
helm/community: hips/hip-00xx-sequencing.md

---
hip: 9999
title: "Better Support for Resource Creation Sequencing"
authors: [ "Joe Beck <joebeck5705@gmail.com>", "Evans Mungai <mbuevans@gmail.com>" ]
created: "2024-11-15"
type: "feature"
status: "draft"
---

## Abstract

> This HIP is to propose a new featureset in Helm 4 to provide Application developers, who create heam charts for their applications, a well supported way of defining what order chart resources and chart dependencies should be deployed to Kubernetes.

Reviewer: Does this proposal also include sequencing for uninstall, or only install/upgrade? For example, waiting to uninstall a resource until all that depends on it have completely uninstalled?

Author: Good point I didn't think about that. At least with uninstall it can just be done backwards to the installation, but upgrade would add some complexity that I'm not sure how to tackle.

Author: I added a section to clarify that this would apply to all 3, installation/uninstall/upgrade

Reviewer 2: and rollback, I assume

## Motivation

Today, to accomplish resource sequencing you have two options. The first is leveraging helm hooks, the second is building required sequencing into the application (via startup code or init containers). The existing hooks and weights can be tedious to build and maintain for Application developers, and built-in app sequencing can unecessarily increase complexity of a Helm application that needs to be maintained by Application Developers. Helm as a package manager should be capable of enabling developers to properly sequence how their applications are deployed, and by providing such a mechanism to developers, this will significantly improve the Application developer's experience.

> Additionally, Helm currently doesn't provide a way to sequence when chart dependencies are deployed, and this featureset would ideally address this.

Reviewer 3: When it comes to sequencing, I'm not sure what you mean is clear in the document. I could come to two different conclusions:

1. The ordering of resources when they are uploaded to the k8s API server in one push
2. An ordered set of communications to the k8s API server where set it sent and completed prior to the next one being sent

Can you please provide some more clarity on what you mean.

Author: 2 is what we're going for here. I'll make a note to try and make that more clear.

Author: Clarified what "resource sequencing" meant in the HIP

## Rationale

> ### Proposal 1: Named Dependencies

Reviewer: Personally, I quite like the general direction of proposal 1—as a replacement of numeric weights entirely—over proposal 2, which extends that. My point of view is I would like to move the numeric weighting into the "alternatives considered" section. But would like to hear other maintainers and community feedback on this.

I have wanted to introduce a depends_on functionality for helm resources—as a dynamic alternative to weights—to allow users to specify a specific order they are applied (a similar concept is used by used by projects like terraform, flux, and docker-compose).

My main reasoning for this is numbered weights are static—you must know the entire list of how all resources should be ordered, and this often takes a lot of finagling and fussing to get it right.

While named dependencies are dynamic—creating a DAG on the fly, based only on when and if certain resources have an actual dependency on one or more other resources.

It might be good to note some of this in the proposal. I could sketch up some suggested text for this if you prefer, but would also be happy for you to take a stab at it in your HIP. What do you think?

Author: Go ahead! I'll add you as a contributor, I appreciate the assistance.

Author: I moved the weights proposal to rejected ideas and tried to capture some of the points you made here in why.

This design was chosen due to simplicity and clarity of how it works for both the Helm developer and Application Operator

### Proposal 2: Weighted Resouces (without hooks)
This design was proposed as an alternative that functions very similarly to hooks today, to provide the same sequencing functionality outside of the scope of hooks.

## Specification

### Proposal 1: Named Dependencies
At a high level, allow Chart Developers to assign named dependency annotations to both their Helm templated resources and Helm chart dependencies that Helm then uses to generate a deployment sequence at installation time. Three new annotations would be added to enable this described next.

- `helm.sh/layer`: Declare a layer that a given resource belongs to. Any numebr of resources can belong to a layer.
- `helm.sh/depends-on/layer`: Declare a layer that must exist and in a ready state before this resource can be deployed.

> - `helm.sh/depends-on/chart`: Declare a chart dependency that must exist and in a ready state before this resource can be deployed. For the chart to be declared ready, all of its resources, with their sequencing order taken into consideration, would need to be deployed and declared ready.

Reviewer: To be clear, `helm.sh/depends-on/chart` refers to some chart that is not listed as a dependency of the current chart? Is that right?
- assuming this because the wait flag should already check that dependent charts are fully installed before installing the current chart
- is the idea that this is limited to other charts that have already been deployed to the same namespace, or is there an idea to check charts in other namespaces (assuming the user has correct RBAC permissions)?

My other question here is, do we want to allow depending on other named k8s resources that are not defined by a chart? Eg, `helm.sh/depends-on/resource`? Or is the idea that users could annotate other resources not defined by a chart with `helm.sh/layer`, declaring that they are considered part of the same layer of a chart?

Author: `helm.sh/depends-on/chart` is referring to any direct chart dependencies. I was unsure if Helm waits to install the current chart until all dependent charts are installed and assumed not. If Helm is already waiting for dependent charts to be installed, then this section can probably be removed?

Reviewer 4: > assuming this because the wait flag should already check that dependent charts are fully installed before installing the current chart

`helm.sh/depends-on/chart` meant for both dependencies and subcharts, will wait regardless of `--wait/--wait-for-jobs` CLI flags being present. Unless we want to have a way to opt-out from sequencing, which we can then assume if `--wait/--wait-for-jobs` are absent, not sequencing. Today, `--wait` will wait for certain resources only. In the proposal we have `helm.sh/resource-ready` which a chart developer can use to define what they mean by "readiness". 

I think we need to add a section explaining how `--wait/--wait-for-jobs` will work with sequencing annotations. I added it to **Open Issues**

> My other question here is, do we want to allow depending on other named k8s resources that are not defined by a chart? Eg, helm.sh/depends-on/resource? Or is the idea that users could annotate other resources not defined by a chart with helm.sh/layer, declaring that they are considered part of the same layer of a chart?

I think using `helm.sh/depends-on/resource` to specify a resource makes sense. Helm would eventually interpret layers as a group of resources to wait for

Reviewer 4: A suggestion came my way regarding reasoning around sequencing chart dependencies/subcharts. Chart authors would want to place such dependencies on a chart-level definition rather than inside chart resources. It makes sense because only one reference of the chart dependency is needed. Any other references are repetition since helm will wait for the chart before deploying the parent chart. We came up with something like below

In the parent `Chart.yaml`
```yaml
name: MyChart
annotations:
  # List of charts that need to be installed before MyChart
  helm.sh/depends-on/charts: "subchart, minio, rabbitmq"
dependencies:
- name: rabbitmq
  version: "1.2.3"
- name: minio
  version: "5.0.0"
```

             ||   [someChartDependency]
             ||     ||
             \/     ||
layer2:    [bar]    ||
              \\   //
               \/ \/
layer3:       [fizz]
```

This approach is prone to circular dependencies. During the templating phase, Helm will have logic to detect, and report any circular dependencies found in the chart templates.

### Proposal 2: Weighted Resouces (without hooks)

Expand on the the weight capabilities of hooks, but allow them to be defined outside of the context of hooks for the chart. Add a new annotation `helm.sh/weight` that helm then uses to order when resources are deployed based on the same weighting rules that hooks use today.

#### Example:
```yaml
# resource 1
metadata:
  name: foo
  annotations:
    helm.sh/weight: "-2"
---
# resource 2
metadata:
  name: bar
  annotations:
    helm.sh/weight: "-1"
---
# resource 3
metadata:
  name: fizz
  annotations:
    helm.sh/weight: 0
```
In this example, Helm would be responsible for resolving the annotations on these three resources and deploy all resources in the following order:

```
-2:    [foo]
        ||
        \/
-1:    [bar]
        ||
        \/
0:     [fizz]
```

### Readiness

> In order to enforce sequencing, new `helm.sh/resource-ready` annotation would be used to determine when a resource is ready, allowing helm to proceed deploying the next group of resources, or failing a deployment. Helm would query, using jsonpath syntax, status fields of the annotated resource. Some native kubernetes resources that have stable APIs e.g `v1` resources such as `Pod` and `Deployment`, would have default queries which can be overriden. In Helm cannot determine how to check a resource's readiness when it should, it will do nothing and log this.

Reviewer: guessing you meant "if" Helm can not determine readiness?
```suggestion
In order to enforce sequencing, new `helm.sh/resource-ready` annotation would be used to determine when a resource is ready, allowing helm to proceed deploying the next group of resources, or failing a deployment. Helm would query, using jsonpath syntax, status fields of the annotated resource. Some native kubernetes resources that have stable APIs e.g `v1` resources such as `Pod` and `Deployment`, would have default queries which can be overriden. If Helm cannot determine how to check a resource's readiness when it should, it will do nothing and log this.
```

#### Example
```yaml
kind: Job
metadata:
  name: barz
  annotations:

>     helm.sh/resource-ready: "status.succeeded==1"

Reviewer: Is the idea that this would be a check that Helm could use to override—kstatus ready condition (as an example implementation)? https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/README.md#the-ready-condition

Author: Yeah

status:
  succeeded: 1
```

#### Sequencing order

Resources with sequencing annotations would be deployed first followed by resources without. If a resource has a `helm.sh/depends-on/chart` annotation, all resources of the referred subchart would be deployed and checked for readiness before deploying the dependant resource. Only resources that Helm can determine their readiness for will be checked. Chart authors would need to annotate their chart resources accordingly. Helm will use default readiness checks for subcharts that do have their templates annotated.

Helm would scope each subchart layer annotation names using a delimiter e.g `someChartDependency#mylayer` to avoid any name collisions. This is an internal implementation detail rather than feature chart authors or operators would need to know.

> `helm template` would print all resources in the order they would be deployed. Groups of resources in a layer would be delimited using a `## Layer: <name>` comment indicate the beginning of each layer

Reviewer 3: The `#` at the front denotes a comment. I'm not sure the k8s yaml parser, that we currently use and is supported by the k8s project, provides a means to parse and use data in comments. This needs to be investigated before this possibility could be considered. See https://github.com/kubernetes-sigs/yaml

Reviewer 4: For reference, the `## Layer: <name>` would be above/below `# Source:`. Maybe there is no need for 2 `#`. We can have it be `# Layer: <name>`.

```yaml
---
# Source: foo/templates/tests/test-connection.yaml
apiVersion: v1
kind: Pod
metadata:
```

---
hip: 9999
title: "Better Support for Resource Creation Sequencing"
authors: [ "Joe Beck <joebeck5705@gmail.com>", "Evans Mungai <mbuevans@gmail.com>" ]
created: "2024-11-15"
type: "feature"
status: "draft"
---

## Abstract

> This HIP is to propose a new featureset in Chart v3, to provide application distributors, who create helm charts for applications, a well supported way of defining what order chart resources and chart dependencies should be deployed to Kubernetes. Helm deploys all manifests at the same time. This HIP will propose a way for Helm to deploy manifests in batches, and inspect states of these resources to enforce sequencing.

Reviewer: Thanks @sabre1041 — I can answer this one.

I suggested "Application Distributor" because Helm [User Profiles](https://github.com/helm/community/blob/main/user-profiles.md) doc calls chart authors "Application Distributors."

Do you think it would be less confusing if we link to that user profile, and not emphasize "application" again afterwards. Like this change?

```suggestion
This HIP is to propose a new featureset in Chart v3, to provide chart authors ([Application Distributor](https://github.com/helm/community/blob/main/user-profiles.md#2-application-distributor) Helm user profile), a well supported way of defining what order chart resources and chart dependencies should be deployed to Kubernetes. Helm deploys all manifests at the same time. This HIP will propose a way for Helm to deploy manifests in batches, and inspect states of these resources to enforce sequencing.
```

> This HIP is to propose a new featureset in Chart v3, to provide application distributors, who create helm charts for applications, a well supported way of defining what order chart resources and chart dependencies should be deployed to Kubernetes. Helm deploys all manifests at the same time. This HIP will propose a way for Helm to deploy manifests in batches, and inspect states of these resources to enforce sequencing.

Reviewer 5: > Thanks @sabre1041 — I can answer this one.
> 
> I suggested "Application Distributor" because Helm [User Profiles](https://github.com/helm/community/blob/main/user-profiles.md) doc calls chart authors "Application Distributors."
> 
> Do you think it would be less confusing if we link to that user profile, and not emphasize "application" again afterwards. Like this change?

Yes that would be very helpful

Additionally, Helm currently doesn't provide a way to sequence when chart dependencies are deployed and this featureset would ideally address this.

## Rationale

This design was chosen due to simplicity and clarity of how it works for both the Helm developer and Application Operator

## Specification

At a high level, allow Chart Developers to assign named dependencies to both their Helm templated resources and Helm chart dependencies that Helm then uses to generate a deployment sequence at installation time.

For Helm CLI, the `--wait=ordered` flag will enable sequencing where resources are applied in groups of layers. SDK users will also be able to enable sequencing by setting a `Wait` boolean flag. Without this flag being enabled, resources will all be applied in one go which is the same behaviour in Chart v2.

Each release will store information of whether sequencing was used or not. This information is used when performing uninstalls and rollbacks.

The following annotations would be added to enable this.

*Additions to templates*

> - `helm.sh/layer`: Annotation to declare a layer that a given resource belongs to. Any number of resources can belong to a layer. A resource can only belong to one layer.

Reviewer 4: > `layer` is not very descriptive into how its being used. End users may not understand how it can/should be used

I'll rename this to `resource-group`. I think this is a clearer name. Its come up before.

> - `helm.sh/layer`: Annotation to declare a layer that a given resource belongs to. Any number of resources can belong to a layer. A resource can only belong to one layer.

Reviewer 4: @sabre1041 here are the renaming changes

https://github.com/helm/community/pull/373/commits/43713e6eee7a398482bc2f13820d904d1a9887b8

- `helm.sh/depends-on/layers`: Annotation to declare layers that must exist and in a ready state before this resource can be deployed. Order of the layers has no significance

*Additions to Chart.yaml*

> - `helm.sh/depends-on/subcharts`: Annotation added to `Chart.yaml` to declare chart dependencies, by name or tags, that must be deployed fully and in a ready state before the chart can be deployed. For the dependency or subchart chart to be declared ready, all of its resources, with their sequencing order taken into consideration, would need to be deployed and declared ready. Order of the charts has no significance.

Reviewer 4: Good catch. I believe this ought to have been "name" or "alias". Alias would be preferred.

> - `helm.sh/depends-on/subcharts`: Annotation added to `Chart.yaml` to declare chart dependencies, by name or tags, that must be deployed fully and in a ready state before the chart can be deployed. For the dependency or subchart chart to be declared ready, all of its resources, with their sequencing order taken into consideration, would need to be deployed and declared ready. Order of the charts has no significance.

Reviewer: Yes good note about alias. I believe this would be caught in the implementation but glad it's called out in the HIP.

- `depends-on`: A new field added to `Chart.yaml` `dependencies` fields that is meant to declare a list of subcharts, by name or tag, that need to be ready before the subchart in questions get installed. This will be used to create a dependency graph for subcharts.

> The installation process would group resources in the same layer and send them to the K8s API Server in one bundle, and once all resources are "ready", the next layer would be installed. A layer would not be considered "ready" and the next layer installed until all resources in that layer are considered "ready". Readiness is described in a later section. A similar process would apply for upgrades. Uninstalls would function on the same layer order, but backwards, where a layer is not uninstalled until all layers that depend on it are first uninstalled. Upgrades would follow the same order as installation.

Reviewer 4: > How are resources ordered as part of a layer? Alphabetically?

Resources in a layer will be sorted the same way Helm 3 does today

https://github.com/helm/helm/blob/9aefd2b32809960203d0178c89522f6bf3d4af9f/pkg/release/util/kind_sorter.go#L31

                             \/ \/
"app" layer:                 [my-app]
```

#### Chart dependencies example

```yaml
name: foo
annotations:
  helm.sh/depends-on/subcharts: ["bar", "rabbitmq"]
dependencies:
  - name: nginx
    version: "18.3.1"
    repository: "oci://registry-1.docker.io/bitnamicharts"
  - name: rabbitmq
    version: "9.3.1"
    repository: "oci://registry-1.docker.io/bitnamicharts"
  - name: bar   # This is a subchart packaged with "foo" in charts dir. It's not pulled from a remote location
    version: "0.1.0"
    depends-on: ["nginx", "rabbitmq"]
    condition: bar.enabled
```

```
 [nginx] [rabbitmq]
    ||       ||
    ||       ||
    \/       \/
     \\     //
      \/   \/
       [bar]
        ||
        ||
        \/
       [foo]
```

In this example, helm will first install and wait for all resources of `nginx` and `rabbitmq` dependencies to be "ready" before attempting to install `bar` resources. Once all resources of `bar` are "ready" then and only then will `foo` chart resources be installed. `foo` would require `rabbitmq` to be ready but since the subchart resources would have been installed before `bar`, this requirement would have been fulfilled.

> This approach of building a directed acyclic graph (DAG) is prone to circular dependencies. During the templating phase, Helm will have logic to detect, and report any circular dependencies found in the chart templates. Helm will also provide a command to print the DAG for development and troubleshooting purposes.

Reviewer 4: > Will non existent layers result in an error or a warning when constructing the DAG?

This would be sent to KAS without waiting for their statuses. All add a "deployment flow" section that will represent this clearly

> This approach of building a directed acyclic graph (DAG) is prone to circular dependencies. During the templating phase, Helm will have logic to detect, and report any circular dependencies found in the chart templates. Helm will also provide a command to print the DAG for development and troubleshooting purposes.

Reviewer: > Will non existent layers result in an error or a warning when constructing the DAG?

@sabre1041 can you clarify what you mean here by "nonexistent layer"? (now called [resource-group](https://github.com/helm/community/pull/373#discussion_r2091721804))

I was guessing you might mean:
- if no resources are part of a `resource-group` that another resource references (with a `depends-on` annotation)?
- or the other way around—if resources are part of a `resource-group` that no other resource depends on?
- or do you mean how would we handle resources that are not part of any `resource-group`?

> This approach of building a directed acyclic graph (DAG) is prone to circular dependencies. During the templating phase, Helm will have logic to detect, and report any circular dependencies found in the chart templates. Helm will also provide a command to print the DAG for development and troubleshooting purposes.

Reviewer 4: @sabre1041 I added a section defining the sequencing workflow Helm would follow. This ought to answer your question.

https://github.com/helm/community/pull/373/commits/4971d775b95efd7d68a151be6a53b7da05d1f700

### Readiness

> In order to enforce sequencing, Helm will check that resources are ready using `kstatus` [ready condition](https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/README.md#the-ready-condition) allowing helm to proceed installing the next group of resources, or fail the install. Users can override checks using optional `helm.sh/readiness-failure` and `helm.sh/readiness-success` annotations which are jsonpath queries to check status fields of the annotated resource. If any of the checks in the lists evaluate as true, the readiness check is determined as successful/failed. `helm.sh/readiness-timeout` annotation can be used to override how long Helm should wait when performing readiness checks before bailing out. It defaults to `10s`.

Reviewer 5: ```suggestion
In order to enforce sequencing, Helm will check that resources are ready using `kstatus` [ready condition](https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/README.md#the-ready-condition) allowing helm to proceed installing the next group of resources, or fail the install. Users can override checks using optional `helm.sh/readiness-failure` and `helm.sh/readiness-success` annotations which are jsonpath queries to check status fields of the annotated resource. If any of the checks in the lists evaluate as true, the readiness check is determined as successful/failed. `helm.sh/readiness-timeout` annotation can be used to override how long Helm should wait when performing readiness checks before raising an error. It defaults to `10s`.
```

> In order to enforce sequencing, Helm will check that resources are ready using `kstatus` [ready condition](https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/README.md#the-ready-condition) allowing helm to proceed installing the next group of resources, or fail the install. Users can override checks using optional `helm.sh/readiness-failure` and `helm.sh/readiness-success` annotations which are jsonpath queries to check status fields of the annotated resource. If any of the checks in the lists evaluate as true, the readiness check is determined as successful/failed. `helm.sh/readiness-timeout` annotation can be used to override how long Helm should wait when performing readiness checks before bailing out. It defaults to `10s`.

Reviewer 5: Does there need to be some commentary/discussion regarding how to construct `readiness-success` and/or `readiness-failure`?

Reviewer 4: > Does there need to be some commentary/discussion regarding how to construct `readiness-success` and/or `readiness-failure`?

@sabre1041 I update the doc to better explain how readiness checks work. https://github.com/helm/community/pull/373/commits/707dd264f246fc19c7c66518cbf8331128d5aa65

> In order to enforce sequencing, Helm will check that resources are ready using `kstatus` [ready condition](https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/README.md#the-ready-condition) allowing helm to proceed installing the next group of resources, or fail the install. Users can override checks using optional `helm.sh/readiness-failure` and `helm.sh/readiness-success` annotations which are jsonpath queries to check status fields of the annotated resource. If any of the checks in the lists evaluate as true, the readiness check is determined as successful/failed. `helm.sh/readiness-timeout` annotation can be used to override how long Helm should wait when performing readiness checks before bailing out. It defaults to `10s`.

Reviewer 4: `helm.sh/readiness-timeout` is a per resource timeout override. Are you suggesting defining a timeout at a higher level, say `30s` for all `Job`s etc?

> In order to enforce sequencing, Helm will check that resources are ready using `kstatus` [ready condition](https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/README.md#the-ready-condition) allowing helm to proceed installing the next group of resources, or fail the install. Users can override checks using optional `helm.sh/readiness-failure` and `helm.sh/readiness-success` annotations which are jsonpath queries to check status fields of the annotated resource. If any of the checks in the lists evaluate as true, the readiness check is determined as successful/failed. `helm.sh/readiness-timeout` annotation can be used to override how long Helm should wait when performing readiness checks before bailing out. It defaults to `10s`.

Reviewer 4: Here is a commit to replace using `helm.sh/readiness-timeout` per resource annotation, in favour of `--readiness-timeout` CLI flag that applies to all resource types. Per resource granularity is unnecessary given that this timeout just leads to a failed release.

https://github.com/helm/community/pull/373/commits/05616838a6b2b85c14f6f2dcae9aa6286ee518da

#### Example
```yaml
kind: Job
metadata:
  name: barz
  annotations:
    helm.sh/readiness-success: ["succeeded==1"] # status fields
    helm.sh/readiness-failure: ["failed==1"] # status fields
    helm.sh/readiness-timeout: 20s # timeout
status:
  succeeded: 1
```

A resources readiness is checked if there is a resource that depends on it as per the sequencing DAG.

#### Sequencing order

> Resources with sequencing annotations in a chart would be deployed first followed by resources without. If the chart has a `helm.sh/depends-on/subcharts` annotation in `Chart.yaml`, all resources of the defined subcharts would be deployed before deploying the main chart. If any sequencing annotations are defined in the subchart resources, Helm will enforce ordering of resources within. Sequencing of resources in a chart are sandboxed within the chart. Sequencing annotations will not affect resources in other charts.

Reviewer 5: ```suggestion
Resources with sequencing annotations in a chart would be deployed first followed by resources without. If the chart has a `helm.sh/depends-on/subcharts` annotation in the `Chart.yaml`, all resources of the defined subcharts would be deployed before deploying the main chart. If any sequencing annotations are defined in the subchart resources, Helm will enforce ordering of resources within. Sequencing of resources in a chart are sandboxed within the chart. Sequencing annotations will not affect resources in other charts.
```

- `helm template` would print all resources in the order they would be deployed. Groups of resources in a layer would be delimited using a `## START layer: <chart> <layer-name>` comment indicating the beginning of each layer and `END layer: <chart> <layer-name>`.

```yaml
## START layer: foo layer1
# resource 1
metadata:
  name: foo
  annotations:
    helm.sh/layer: layer1
---
# resource 2
metadata:
  name: bar
  annotations:
    helm.sh/layer: layer1
## END layer: foo layer1
---
## START layer: bar layer1
# resource 3
metadata:
  name: fizz
  annotations:
    helm.sh/layer: layer2
    helm.sh/depends-on/layers: "layer1"
## END layer: bar layer2
```

## Backwards compatibility

Helm will continue to install/upgrade/uninstall/rollback all resources and dependencies at one go for all charts using `Charts v2` and below.

## Security implications

None.

## How to teach this

> - Document how sequencing works in the official helm documentation website. Include ordering of kubernetes resources that Helm enforces when applying resources to the cluster. Examples will be added to best demonstrate how this feature works.

Reviewer 5: ```suggestion
- Document how sequencing works in the official helm documentation website. Include ordering of Kubernetes resources that Helm enforces when applying resources to the cluster. Examples will be added to best demonstrate how this feature works.
```

> - Document how to this feature works for SDK users.

Reviewer 5: ```suggestion
- Document how this feature works for SDK users.
```
